Add OPL tutorial (#423)
oasisprotocol/docs: docs/dapp/opl/enclave.md

# Ballot Contract

We will next write the smart contract that holds private data and will be

> deployed to the Oasis Privacy Layer.

Reviewer: Probably Sapphire instead of OPL?

```suggestion
deployed to the confidential chain such as Oasis Sapphire.
```

Create a new Solidity contract named `BallotBoxV1.sol`.

> ```solidity

Reviewer: ```suggestion

```solidity
```

            }
        }
        postMessage("ballotClosed", abi.encode(_proposalId, topChoice));
        emit BallotClosed(_proposalId, topChoice);
        delete _ballots[_proposalId];
    }
}
```

#### Autoswitch

Note that we will be switching to the Binance Smart Chain as a default when
we deploy to testnet.
```solidity
    constructor(address dao) Enclave(dao, autoswitch("bsc")) {
        registerEndpoint("createBallot", _oplCreateBallot);
    }
```

#### Event

Closing a ballot has implications on the host chain network.

```solidity
    function _closeBallot(ProposalId _proposalId, Ballot storage _ballot) internal {
        uint256 topChoice;
        uint256 topChoiceCount;
        for (uint8 i; i < _ballot.params.numChoices; ++i) {
            uint256 choiceVoteCount = _ballot.voteCounts[i] & (type(uint256).max >> 1);
            if (choiceVoteCount > topChoiceCount) {
                topChoice = i;
                topChoiceCount = choiceVoteCount;
            }
        }
        postMessage("ballotClosed", abi.encode(_proposalId, topChoice));
        emit BallotClosed(_proposalId, topChoice);
        delete _ballots[_proposalId];
    }
```

#### Private

> Private variable `_ballots` is encrypted.

Reviewer: ```suggestion
Private variable `_ballots` is encrypted on Sapphire.
```

# Ballot Contract

> We will next write the smart contract that holds private data and will be

Reviewer: ```suggestion
Next, we will write the smart contract that holds private data. This smart contract will be executed inside a trusted execution environment (TEE) on the Oasis Sapphire chain,
```

> deployed to a confidential chain such as Oasis Sapphire.

Reviewer: ```suggestion
hence we will call it the *enclave* smart contract.
```

        return _ballots[id].active;
    }

    function _oplCreateBallot(bytes calldata args) internal returns (Result) {
        (ProposalId id, ProposalParams memory params) = abi.decode(
            args,
            (ProposalId, ProposalParams)
        );
        Ballot storage ballot = _ballots[id];
        ballot.params = params;
        ballot.active = true;
        for (uint256 i; i < params.numChoices; ++i) ballot.voteCounts[i] = 1 << 255; // gas usage side-channel resistance.
        return Result.Success;
    }

    function _closeBallot(ProposalId _proposalId, Ballot storage _ballot) internal {
        uint256 topChoice;
        uint256 topChoiceCount;
        for (uint8 i; i < _ballot.params.numChoices; ++i) {
            uint256 choiceVoteCount = _ballot.voteCounts[i] & (type(uint256).max >> 1);
            if (choiceVoteCount > topChoiceCount) {
                topChoice = i;
                topChoiceCount = choiceVoteCount;
            }
        }
        postMessage("ballotClosed", abi.encode(_proposalId, topChoice));
        emit BallotClosed(_proposalId, topChoice);
        delete _ballots[_proposalId];
    }
}
```

#### Autoswitch

> Note that we will be switching to the Binance Smart Chain as a default when

Reviewer: ```suggestion
In this tutorial the enclave contract will talk to the host samrt contract deployed on the Binance Smart Chain (`bsc`).
```

> we deploy to testnet.

Reviewer: ```suggestion
:::tip

Autoswitch will automatically pick the Testnet host network, if the enclave network is also Testnet.

:::

Autoswitch supports the following networks:

- `ethereum`
- `goerli`
- `optimism`
- `bsc`
- `bsc-testnet`
- `polygon`
- `fantom`
- `fantom-testnet`
- `moonriver`
- `arbitrum-one`
- `arbitrum-nova`
- `sapphire`
- `sapphire-testnet`
- `polygon-mumbai`
- `avalanche`
- `avalanche-fuji`
- `arbitrum-testnet`
```

Author: Great suggestion.

```solidity
    constructor(address dao) Enclave(dao, autoswitch("bsc")) {
        registerEndpoint("createBallot", _oplCreateBallot);
    }
```

#### Event

> Closing a ballot has implications on the host chain network.

Reviewer: ```suggestion
Closing a ballot has implications on the host chain network (`postMessage()`):
```

```solidity
    function _closeBallot(ProposalId _proposalId, Ballot storage _ballot) internal {
        uint256 topChoice;
        uint256 topChoiceCount;
        for (uint8 i; i < _ballot.params.numChoices; ++i) {
            uint256 choiceVoteCount = _ballot.voteCounts[i] & (type(uint256).max >> 1);
            if (choiceVoteCount > topChoiceCount) {
                topChoice = i;
                topChoiceCount = choiceVoteCount;
            }
        }
        postMessage("ballotClosed", abi.encode(_proposalId, topChoice));
        emit BallotClosed(_proposalId, topChoice);
        delete _ballots[_proposalId];
    }
```

#### Private

> Private variable `_ballots` is encrypted on Sapphire.

Reviewer: ```suggestion
The private variable `_ballots` is encrypted on Sapphire.
```
